Adding a tool FixVcfHeader.

nh13 opened the pull request:
This tool replaces or fixes up the VCF header. In the latter case, it
adds dummy header lines for INFO or FORMAT fields found in the records
but not in the input VCF.

nh13 commented:
@yfarjoun This tool should fixup some VCFs that don't play nice with Picard or GATK.
@vdauwera can you comment on the docs; I am not sure I am doing it right based on your recent round of changes.

vdauwera commented on src/java/picard/vcf/FixVcfHeader.java:
```diff
@@ -0,0 +1,199 @@
+)
+public class FixVcfHeader extends CommandLineProgram {
+    static final String USAGE_SUMMARY = "Replaces or fixes a VCF header.";
+    static final String USAGE_DETAILS = "Replaces or fixes a VCF heade.<br />" +
```
Repeating the summary sentence isn't necessary

vdauwera commented on src/java/picard/vcf/FixVcfHeader.java:
```diff
@@ -0,0 +1,199 @@
+    static final String USAGE_SUMMARY = "Replaces or fixes a VCF header.";
+    static final String USAGE_DETAILS = "Replaces or fixes a VCF heade.<br />" +
+            "This tool will either replace the header in the input file with the given VCF header or will try to fixup the input header. " +
+            "In the latter case, two-passes over the input file are required, and any INFO and FORMAT fields found in the VCF " +
```
"two-passes" no need for hyphen

vdauwera replied:
I don't understand the two passes thing -- are we running two separate commands, or are the two passes done internally?

vdauwera commented on src/java/picard/vcf/FixVcfHeader.java:
```diff
@@ -0,0 +1,199 @@
+            "java -jar picard.jar FixVcfHeader \\<br />" +
+            "     I=input.vcf \\<br />" +
+            "     O=fixed.vcf \\<br />" +
+            "     HEADER=header.vcf" +
```
Can HEADER be a full vcf or should it contain only the header that we want to use as base for fixing?

vdauwera commented on src/java/picard/vcf/FixVcfHeader.java:
```diff
@@ -0,0 +1,199 @@
+            "     O=fixed.vcf \\<br />" +
+            "     HEADER=header.vcf" +
+            "</pre>" +
+            "<hr />" +
```
this hr is unnecessary

vdauwera commented:
Minor comments and request for clarification.

nh13 commented:
@yfarjoun I have added the functionality we discussed in #528, and a few more tests.  @vdauwera I have updated the documentation, but it could use another set of eyes.

vdauwera commented on src/main/java/picard/vcf/FixVcfHeader.java:
```diff
@@ -0,0 +1,225 @@
+            "java -jar picard.jar FixVcfHeader \\<br />" +
+            "     I=input.vcf \\<br />" +
+            "     O=fixed.vcf \\<br />" +
+            "     FIX_HEADER=true \\<br />" +
```
doesn't need `\\<br />` since this is the last line of the command

nh13 replied:
Done

vdauwera commented on src/main/java/picard/vcf/FixVcfHeader.java:
```diff
@@ -0,0 +1,225 @@
+    static final String USAGE_SUMMARY = "Replaces or fixes a VCF header.";
+    static final String USAGE_DETAILS =
+            "This tool will either replace the header in the input VCF file (INPUT) with the given VCF header (HEADER) or will try to fixup " +
+            "the input header by examining the variants in the input VCF file (INPUT).  In the latter case, this tool will perform two passes " +
```
"try to fixup" -> attempt to fill in any field definitions that are missing in

nh13 replied:
Done

vdauwera commented on src/main/java/picard/vcf/FixVcfHeader.java:
```diff
@@ -0,0 +1,225 @@
+            "     FIX_HEADER=false \\<br />" +
+            "     HEADER=header.vcf" +
+            "</pre>" +
+            "<h4>Fix up usage example:</h4>" +
```
Fix up -> Fix header

nh13 replied:
Done

vdauwera commented on src/main/java/picard/vcf/FixVcfHeader.java:
```diff
@@ -0,0 +1,225 @@
+    @Option(shortName=StandardOptionDefinitions.OUTPUT_SHORT_NAME, doc="The output VCF/BCF file.")
+    public File OUTPUT;
+
+    @Option(shortName="F", doc="True if we are to fix the header, false if we are to replace the header (requires HEADER).")
```
Wouldn't it be simpler to get rid of this boolean and just test for whether we have a HEADER input?

nh13 replied:
Done.

vdauwera commented on src/main/java/picard/vcf/FixVcfHeader.java:
```diff
@@ -0,0 +1,225 @@
+    @Option(shortName="N", doc="Check only the first N records when searching for missing INFO and FORMAT fields.", optional=true)
+    public int CHECK_FIRST_N_RECORDS = -1;
+
+    @Option(shortName="H", doc="The replacement VCF header.  Omit to infer from the input.", optional=true)
```
"Omit to infer from the input." is a bit cryptic. I would leave it out entirely -- the argument itself is fairly self-explanatory, especially once you've read the tool description.

nh13 replied:
Done.

vdauwera commented:
Made a few comments/recommendations.

nh13 commented:
Comments addressed and I will merge once the tests pass.
